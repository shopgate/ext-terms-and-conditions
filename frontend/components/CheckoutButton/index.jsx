import React from 'react';
import PropTypes from 'prop-types';
import I18n from '@shopgate/pwa-common/components/I18n';
import Link from '@shopgate/pwa-common/components/Link';
import RippleButton from '@shopgate/pwa-ui-shared/RippleButton';
import { CHECKOUT_PATH } from '@shopgate/pwa-common/constants/RoutePaths';
import connect from './connector';
import styles from './style';

/**
 * CheckoutButton renders button
 * @param {Props} props props.
 * @returns {JSX}
 */
const CheckoutButton = props => (
  <Link href={CHECKOUT_PATH} disabled={!props.isActive || !props.isOrderable}>
    <RippleButton
      disabled={!props.isActive || !props.isOrderable}
      type="regular"
      className={props.isActive ? styles.button : styles.disabledButton}
    >
      <I18n.Text string="cart.checkout" />
    </RippleButton>
  </Link>
);

CheckoutButton.propTypes = {
  isActive: PropTypes.bool,
  isOrderable: PropTypes.bool,
};

CheckoutButton.defaultProps = {
  isActive: true,
  isOrderable: false,
};

export default connect(CheckoutButton);
