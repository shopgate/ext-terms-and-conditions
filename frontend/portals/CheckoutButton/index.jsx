import React from 'react';
import PropTypes from 'prop-types';
import DisabledCheckoutButton from '../../components/DisabledCheckoutButton';
import connect from './connector';

/**
 * @param {Object} props props.
 * @returns {JSX}
 */
const CheckoutButton = (props) => {
  if (!props.isOrderable) {
    return (
      <DisabledCheckoutButton
        isOrderable={props.isOrderable}
        showTermsNotice={props.showTermsNotice}
      />
    );
  }

  return props.children;
};

CheckoutButton.propTypes = {
  children: PropTypes.node.isRequired,
  isOrderable: PropTypes.bool,
  showTermsNotice: PropTypes.func,
};

CheckoutButton.defaultProps = {
  isOrderable: false,
  showTermsNotice: () => {},
};
export default connect(CheckoutButton);
