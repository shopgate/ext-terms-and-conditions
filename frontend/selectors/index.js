import { createSelector } from 'reselect';
import { getCartProducts } from '@shopgate/pwa-common-commerce/cart/selectors';
import { getProducts } from '@shopgate/pwa-common-commerce/product/selectors/product';
import getConfig from '../helpers/getConfig';

const { checkboxValues } = getConfig();

/**
 * @param {Object} state state
 * @returns {Object}
 */
const getExtensionsState = state => state.extensions;

const getExtensionState = createSelector(
  getExtensionsState,
  extensionsState => extensionsState['@shopgate/terms-and-conditions/isOrderableByTerms']
);

export const getTermsStatus = createSelector(
  getExtensionState,
  extensionState => extensionState.isOrderable
);

export const getCartProductIds = createSelector(
  getCartProducts,
  (cartProducts) => {
    if (!cartProducts) {
      return null;
    }
    return cartProducts.map(cartProduct => cartProduct.product.id);
  }
);

export const getTermProductIds = createSelector(
  getCartProductIds,
  getProducts,
  (cartProductIds, products) => {
    const productsToCheck = [];
    if (!cartProductIds) {
      return null;
    }
    cartProductIds.forEach((cartProductId) => {
      if (products.hasOwnProperty(cartProductId)) {
        const testBaseProductId = products[cartProductId].productData.baseProductId;
        const idToPush = (testBaseProductId || cartProductId);
        productsToCheck.push(idToPush);
      }
    });
    return productsToCheck;
  }
);

export const getTermsToDisplay = createSelector(
  getTermProductIds,
  (productIds) => {
    if (!productIds) {
      return null;
    }
    return checkboxValues.filter((value =>
      productIds.includes(value.displayOn) || value.displayOn === 'all products'));
  }
);

export const getCheckValues = createSelector(
  getTermsToDisplay,
  (terms) => {
    const checkedValues = [];
    if (!terms) {
      return null;
    }
    terms.forEach((product, index) => {
      checkedValues.push({
        displayFor: `${product.displayOn}`,
        checkedValues: [],
      });
      product.productCheckboxValues.forEach(() => {
        checkedValues[index].checkedValues.push({ checked: false });
      });
    });
    return checkedValues;
  }
);

