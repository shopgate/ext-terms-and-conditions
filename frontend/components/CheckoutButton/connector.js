import { connect } from 'react-redux';
import { getTermsStatus } from '../../selectors';

/**
 * Maps the terms and conditions is orderable State.
 * @param {Object} state The current application state.
 * @returns {Object} The extended component props.
 */
const mapStateToProps = state => ({
  isOrderable: getTermsStatus(state),
});

export default connect(mapStateToProps);
