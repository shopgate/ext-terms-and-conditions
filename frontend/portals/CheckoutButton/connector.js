import { connect } from 'react-redux';
import { getTermsStatus } from '../../selectors';
import showTermsNotice from '../../actions';

/**
 * Maps the terms and conditions is orderable State.
 * @param {Object} state The current application state.
 * @returns {Object} The extended component props.
 */
const mapStateToProps = state => ({
  isOrderable: getTermsStatus(state),
});

/**
 * Map showTermsNotice action to the component props.
 * @param {Function} dispatch The redux dispatch function
 * @return {Object} The extended component props.
 */
const mapDispatchToProps = dispatch => ({
  showTermsNotice: () => (
    dispatch(showTermsNotice)
  ),
});

export default connect(mapStateToProps, mapDispatchToProps);
