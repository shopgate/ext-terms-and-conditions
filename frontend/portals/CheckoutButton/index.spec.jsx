import React from 'react';
import { mount } from 'enzyme';
import { createMockStore } from '@shopgate/pwa-common/store';
import { Provider } from 'react-redux';
import { getTermsStatus } from '../../selectors';
import CheckoutButton from './index';

jest.mock('../../components/DisabledCheckoutButton', () => () => <div>Disabled Checkbox</div>);

jest.mock('../../selectors', () => ({
  getTermsStatus: jest.fn().mockReturnValue(false),
}));

const store = createMockStore();

describe('<CheckoutButton />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render as expected when not orderable', () => {
    getTermsStatus.mockReturnValue(false);
    const wrapper = mount((
      <Provider store={store}>
        <CheckoutButton>Child</CheckoutButton>
      </Provider>
    ));
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('div').exists()).toBe(true);
  });

  it('should render props.children when orderable', () => {
    getTermsStatus.mockReturnValue(true);
    const wrapper = mount((
      <Provider store={store}>
        <CheckoutButton>Child</CheckoutButton>
      </Provider>
    ));
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('div').exists()).toBe(false);
  });
});
