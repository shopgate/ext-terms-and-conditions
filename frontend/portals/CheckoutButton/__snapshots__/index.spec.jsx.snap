// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckoutButton /> should render as expected when not orderable 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(CheckoutButton)>
    <CheckoutButton
      isOrderable={false}
      showTermsNotice={[Function]}
    >
      <Component
        isOrderable={false}
        showTermsNotice={[Function]}
      >
        <div>
          Disabled Checkbox
        </div>
      </Component>
    </CheckoutButton>
  </Connect(CheckoutButton)>
</Provider>
`;

exports[`<CheckoutButton /> should render props.children when orderable 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(CheckoutButton)>
    <CheckoutButton
      isOrderable={true}
      showTermsNotice={[Function]}
    >
      Child
    </CheckoutButton>
  </Connect(CheckoutButton)>
</Provider>
`;
